### ISSUE TITLE
Task.memoize doesn't play well with Local

### ISSUE THREAD

**Issue by guymers** [monix/monix#856]
I am trying to use `Local` as a replacement for a `ThreadLocal` in order to implement OpenTracing support (https://github.com/opentracing/opentracing-java/blob/release-0.31.0/opentracing-util/src/main/java/io/opentracing/util/ThreadLocalScopeManager.java). However not all downstream `Task`s see the active `Scope`. Is what I am trying to accomplish a supported use case?

If I make this change everything works as I would expect.
```
diff --git a/monix-eval/shared/src/main/scala/monix/eval/internal/TaskRestartCallback.scala b/monix-eval/shared/src/main/scala/monix/eval/internal/TaskRestartCallback.scala
index 9752c396..6a91b3e0 100644
--- a/monix-eval/shared/src/main/scala/monix/eval/internal/TaskRestartCallback.scala
+++ b/monix-eval/shared/src/main/scala/monix/eval/internal/TaskRestartCallback.scala
@@ -155,8 +155,7 @@ private[internal] object TaskRestartCallback {
     private[this] var previousLocals: Local.Context = _
 
     override protected def prepareStart(task: Task.Async[_]): Unit = {
-      preparedLocals =
-        if (task.restoreLocals) Local.getContext() else null
+      preparedLocals = Local.getContext()
     }
 
     override def prepareCallback: Callback[Throwable, Any] =
```

The below test approximates what an average `Task` stack looks like:
```
val memoizedTask = Task.delay {
  5
}.memoize

val i = AtomicInt(0)

val t = for {
  local <- TaskLocal(0)
  ii <- Task.shift.flatMap(_ => Task.delay {
    i.incrementAndGet()
  })
  _ <- local.write(ii)
  result <- Task.parZip2(
    memoizedTask.flatMap { _ =>
      Task.sleep(100.millis).flatMap { _ =>
        local.read
      }
    },
    memoizedTask.flatMap { _ =>
      Task.sleep(100.millis).flatMap { _ =>
        local.read
      }
    },
  )
} yield {
  println(result)
}

t.runToFutureOpt
t.runToFutureOpt
t.runToFutureOpt
```

Running this will never return `(0,0)`, `(1,1)`, `(2,2)`.

However if you remove the `Task.shift` or remove the `memoize` it does.

**Comment by Avasil** [comment]
We'd like to support such a use cases if it doesn't currently work as expected

What result do you get instead? I get:
```
(1,1)
(2,2)
(3,3)
```

Sometimes printed in different order. It's the same without `memoize` / `shift` for me and seems like expected result to me. Am I missing something?

You could also check current master if it's better/worse (has some changes to `TaskLocal`): `"3.0.0-RC2-SNAPSHOT-55a922a-SNAPSHOT"`?

**Comment by guymers** [comment]
Sorry I would expect `(1,1) (2,2) (3,3)` in any order not `(0,0), (1,1), (2,2)`

That is strange that it works for you. This is a sample of what I get when I run https://github.com/guymers/monix-local-temp

```
List((1,1), (0,0), (3,3))
List((1,1), (2,2), (3,3))
List((0,0), (3,3), (2,2))
List((1,1), (2,2), (3,0))
List((1,1), (0,2), (0,3))
List((1,1), (0,0), (0,0))
List((0,0), (0,2), (3,3))
List((0,1), (0,2), (3,3))
List((1,0), (0,0), (3,3))
List((0,0), (2,2), (3,3))
```

If I remove the call to `memoize` it seems to return the expected every time.

If I remove the call to `shift` it seems to return the expected ~80% of the time. (Only seems to return `List((1,1), (0,0), (0,0))` as the failing case)

This is on a different computer to the previous, same OS however.

**Comment by Avasil** [comment]
Try with `TracingScheduler`:
```scala
implicit val ec: Scheduler = TracingScheduler(monix.execution.Scheduler.Implicits.global)
```
Are you planning to release it later as a library? Might be useful for other users if it works out

**Comment by guymers** [comment]
Using `TracingScheduler` doesn't make a difference. 

You get the expected result every time you run the above? I just ran it on Travis and it seems to also get unexpected results https://travis-ci.org/guymers/monix-local-temp/builds/514106482

I wasn't planning on making a library the only thing required is to copy `ThreadLocalScopeManager` and replace the `ThreadLocal` then copy `ThreadLocalScope`and get it to compile.

**Comment by Avasil** [comment]
Ok, I can reproduce it, I guess I couldn't do it because I was running it through IntelliJ

We'll look into it

**Comment by Avasil** [comment]
Haven't think of proper fix yet but I minimized this case further:

```scala
val memoizedTask = Task.delay {
  5
}.memoize

val t = for {
  local <- TaskLocal(0)
  result <- Task.parZip2(
    local.write(1) >> memoizedTask >> local.read,
    local.write(2) >> memoizedTask >> local.read
  )
  } yield {
    println(result)
  }

t.runToFutureOpt
```
Will print `(1, 1)` or `(2, 2)` 

```scala
val memoizedTask = Task.delay {
  5
}.memoize

val t = for {
  local <- TaskLocal(0)
  _  <- memoizedTask
  result <- Task.parZip2(
    local.write(1) >> memoizedTask >> local.read,
    local.write(2) >> memoizedTask >> local.read
  )
  } yield {
    println(result)
  }

t.runToFutureOpt
```

Will always print `(1, 2)` (as expected)

If we change `Task.memoize` to restoreLocals then it will work in the first example (your situation) but not in the following case:
```scala
val task = for {
  local <- TaskLocal(0)
  v1 <- (local.write(100).flatMap(_ => local.read)).memoize
  _ <- Task.shift
  v2 <- local.read
} yield (v1, v2) // (100, 0)
```

**Comment by Avasil** [comment]
I think this is fixed in #887 but I'm not 100% sure so I'll leave it open for now, we'll see soon. :)

Curren't model shares between tasks in `Task.parZip` so my previous comment is not valid but the original issue by @guymers seems to behave as expected. I can add a test case later

**Comment by guymers** [comment]
Thanks for fixing this, I'll try to find some time to test it against a real world application.

**Comment by Avasil** [comment]
@guymers It was fully fixed in PR that will ship in `3.0.0` (I really hope to release it in a matter of weeks, 3 prs to merge left) but I can release a SNAPSHOT if you're interested

**Comment by guymers** [comment]
If it is easy to do sure. I wont be merging it until there is a released version so publishing local will work for me.

**Comment by Avasil** [comment]
@guymers 
I have released `3.0.0-SNAPSHOT-6a732d6-SNAPSHOT` in sonatype snapshot repository. It should be available soon

Any feedback regarding the latest Local model will be very valuable

**Comment by guymers** [comment]
Got around to doing some basic testing with that snapshot version and my open tracing spans are now correct, cheers.
